scripts/fuzz_opt.py

```diff
@@ -1241,6 +1246,53 @@ def handle(self, wasm):
         compare_between_vms(fix_output(wasm_exec), fix_output(evalled_wasm_exec), 'CtorEval')
 
 
+# Tests wasm-merge
+class Merge(TestCaseHandler):
+    frequency = 0.25
+
+    def handle(self, wasm):
+        # generate a second wasm file to merge. note that we intentionally pick
+        # a smaller size than the main wasm file, so that reduction is
+        # effective (i.e., as we reduce the main wasm to small sizes, we also
+        # end up with small secondary wasms)
+        # TODO: add imports and exports that connect between the two
+        wasm_size = os.stat(wasm).st_size
+        second_size = min(wasm_size, random_size())
+        second_input = abspath('second_input.dat')
+        make_random_input(second_size, second_input)
+        second_wasm = abspath('second.wasm')
+        run([in_bin('wasm-opt'), second_input, '-ttf', '-o', second_wasm] + FUZZ_OPTS + FEATURE_OPTS)
+
+        # sometimes also optimize the second module
+        if random.random() < 0.5:
+            opts = get_random_opts()
+            run([in_bin('wasm-opt'), second_wasm, '-o', second_wasm, '-all'] + FEATURE_OPTS + opts)
+
+        # merge the wasm files. note that we must pass -all, as even if the two
+        # inputs are MVP, the output may have multiple tables and multiple
+        # memories (and we must also do that in the commands later down).
+        merged = abspath('merged.wasm')
+        run([in_bin('wasm-merge'), wasm, 'first', abspath('second.wasm'), 'second', '-o', merged] + FEATURE_OPTS + ['-all'])
+
+        # sometimes also optimize the merged module
+        if random.random() < 0.5:
+            opts = get_random_opts()
+            run([in_bin('wasm-opt'), merged, '-o', merged, '-all'] + FEATURE_OPTS + opts)
+
+        # verify that merging in the second module did not alter the output.
+        output = run_bynterp(wasm, ['--fuzz-exec-before', '-all'])
+        output = fix_output(output)
+        merged_output = run_bynterp(merged, ['--fuzz-exec-before', '-all'])
+        merged_output = fix_output(merged_output)
+        # a complication is that the second module's exports are appended, so we
+        # have extra output. to handle that, just prune the tail, so that we
+        # only compare the original exports from the first module.
+        # TODO: compare the second module's exports to themselves as well, but
+        #       they may have been renamed due to overlaps...
```

Comment:
Silently renaming exports seems bad! Should this be an error by default instead?

Reply:
I don't think this should error, since there are use cases like e.g. two modules export "main" and so after renaming we end up with "main" and "main_1". The user of `wasm-merge` knows that the first is the first module's and so forth, so they can call them at the appropriate times. (The renamings are deterministic, which makes that possible.)

Reply:
In general I wouldn't expect a user to have a detailed enough global understanding of their system to anticipate name conflicts. In particular for the example of `main`, I expect that merges with two exported `main`s are more likely to be accidental than intentional. I agree that it should be possible to let wasm-merge handle resolve conflicts automatically, but I think that should be opt-in and that erroring out would be a safer default.

Also, WDYT about resolving conflicts or even handling all exports by renaming each _export_ from an original _module_ to _module.export_ in the combined module? That would make conflicts impossible by construction.

Reply:
We might want multiple modes here, yeah. I'm not sure what the default should be, though. Renaming all exports to `module.export` would mean we don't handle well the simple use case of a main module being merged with side modules, and the side modules have no final exports to the outside - in that case we just want the main module's exports to be kept as they are, and the side module's exports to be "dropped."

So maybe we want some modes like this:

* Rename conflicting export names.
* Error on conflicting export names.
* Keep exports from one module, drop all the rest. (Or, perhaps a set of modules could keep their exports, with either renaming or erroring.)